From 29df782bead924a63e18f5f431c980d56d6bb073 Mon Sep 17 00:00:00 2001
From: Christian Ehrhardt <christian.ehrhardt@codasip.com>
Date: Thu, 17 Oct 2024 14:23:40 +0200
Subject: [PATCH] cheri: Fix pointer handling in job table

Another instance of a realloc where pointers are translated by
adding an inter object pointer difference.

Upstream-Status: Pending

---
 shell/ash.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/shell/ash.c b/shell/ash.c
index 42d95b839..488411f6c 100644
--- a/shell/ash.c
+++ b/shell/ash.c
@@ -4690,7 +4690,10 @@ growjobtab(void)
 			l -= sizeof(*jp);
 			jq--;
 #define joff(p) ((struct job *)((char *)(p) + l))
-#define jmove(p) (p) = (void *)((char *)(p) + offset)
+#define jmove(p) do { \
+	(p) = (void *)((char *)(p) + offset); \
+	(p) = (void *)jp + ((void *)(p) - (void *)jp); \
+} while (0);
 			if (joff(jp)->ps == &jq->ps0)
 				jmove(joff(jp)->ps);
 			if (joff(jp)->prev_job)
